File: services/api/source/code/entities/videos/video.py
from entities.videos.uploaded_video import UploadedVideo
from pydantic import HttpUrl
from typing import Optional, List, Set, ClassVar
import datetime

# a list of required fields to update while listing the video for the first time
# otherwise, they will remain empty
REQUIRED_FIELDS_ON_LISTING = ['title', 'description']
# a set of fields which are allowed to update by design
ALLOWED_UPDATE_FIELDS_FOR_LISTED_VIDEOS = {'title', 'description', 'is_private'}
ALLOWED_UPDATE_FIELDS_FOR_NEW_LISTING = ALLOWED_UPDATE_FIELDS_FOR_LISTED_VIDEOS.union({'listing_time'})


class Video(UploadedVideo):
    # static
    REQUIRED_FIELDS_ON_LISTING: ClassVar[List[str]] = REQUIRED_FIELDS_ON_LISTING
    ALLOWED_UPDATE_FIELDS_FOR_NEW_LISTING: ClassVar[Set[str]] = ALLOWED_UPDATE_FIELDS_FOR_NEW_LISTING
    ALLOWED_UPDATE_FIELDS_FOR_LISTED_VIDEOS: ClassVar[Set[str]] = ALLOWED_UPDATE_FIELDS_FOR_LISTED_VIDEOS

    # per instance
    title: Optional[str]
    description: Optional[str]
    size_in_bytes: Optional[int]
    duration_seconds: Optional[int]
    video_type: Optional[str]
    thumbnail_url: Optional[HttpUrl]
    is_private: Optional[bool]
    listing_time: Optional[datetime.datetime]
    class Config:
        underscore_attrs_are_private = True
    
    _pagination_index: Optional[int]
    _storage_object_key: Optional[str]
    _storage_thumbnail_key: Optional[str]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        # privates
        self._pagination_index = kwargs.get('pagination_index', None)
        self._storage_object_key = kwargs.get('storage_object_key', None)
        self._storage_thumbnail_key = kwargs.get('storage_thumbnail_key', None)


    @property
    def pagination_index(self) -> str:
        return self._pagination_index


    @property
    def storage_object_key(self) -> str:
        return self._storage_object_key


    @property
    def storage_thumbnail_key(self) -> str:
        return self._storage_thumbnail_key

    
    def is_listed(self) -> bool:
        return self.listing_time is not None
    

    def is_not_listed(self) -> bool:
        return not self.is_listed()

File: services/ui/code/src/app/core/models/entities/videos/video.ts
import { IVideo } from "../../backend/videos.types";
import assertField from "../assert-field";
import UploadedVideo from "./uploaded-video";
import { removeUndefinedValuesKeysFromObject } from '../../../../common/utils';

class Video extends UploadedVideo {
    title: string | null;
    description: string | null;
    sizeInBytes: number;
    durationSeconds: number;
    videoType: string;
    thumbnailUrl: string;
    isPrivate: boolean;
    listingTime: string | null;
    paginationIndex: number;

    constructor(
        hashId: string,
        userId: string,
        uploadTime: string,
        title: string | undefined,
        description: string | undefined,
        sizeInBytes: number,
        durationSeconds: number,
        videoType: string,
        thumbnailUrl: string,
        isPrivate: boolean,
        listingTime: string | undefined,
        paginationIndex: number
    ) {
        super(hashId, userId, uploadTime);

        this.title = title || null;
        this.description = description || null;
        this.listingTime = listingTime || null;

        assertField(this.constructor.name, 'sizeInBytes', sizeInBytes);
        this.sizeInBytes = sizeInBytes;
        assertField(this.constructor.name, 'durationSeconds', durationSeconds);
        this.durationSeconds = durationSeconds;
        assertField(this.constructor.name, 'videoType', videoType);
        this.videoType = videoType;
        assertField(this.constructor.name, 'thumbnailUrl', thumbnailUrl);
        this.thumbnailUrl = thumbnailUrl;
        assertField(this.constructor.name, 'isPrivate', isPrivate);
        this.isPrivate = isPrivate;
        assertField(this.constructor.name, 'paginationIndex', paginationIndex);
        this.paginationIndex = paginationIndex;
    }

    public isListed(): boolean {
        return this.listingTime !== null;
    }

    static override fromInterface(source: IVideo): Video {
        return new Video(
            source.hash_id,
            source.user_id,
            source.upload_time,
            source.title,
            source.description,
            source.size_in_bytes,
            source.duration_seconds,
            source.video_type,
            source.thumbnail_url,
            source.is_private,
            source.listing_time || undefined,
            source.pagination_index
        )
    }

    public override toInterface(): IVideo {
        const result = {
            ...super.toInterface(),
            title: this.title || undefined,
            description: this.description || undefined,
            size_in_bytes: this.sizeInBytes,
            duration_seconds: this.durationSeconds,
            video_type: this.videoType,
            thumbnail_url: this.thumbnailUrl,
            is_private: this.isPrivate,
            listing_time: this.listingTime || undefined,
            pagination_index: this.paginationIndex
        }
        removeUndefinedValuesKeysFromObject(result);
        return result;
    }
}



export default Video;